src/main/java/hudson/plugins/mercurial/traits/MercurialInstallationSCMSourceTrait.java

```diff
@@ -0,0 +1,157 @@
+import hudson.plugins.mercurial.MercurialSCMBuilder;
+import hudson.plugins.mercurial.MercurialSCMSource;
+import hudson.plugins.mercurial.MercurialSCMSourceContext;
+import hudson.scm.SCM;
+import hudson.util.ListBoxModel;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.trait.SCMBuilder;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceRequest;
+import jenkins.scm.api.trait.SCMSourceTrait;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+/**
+ * Exposes {@link MercurialInstallation} configuration of a {@link MercurialSCMSource} as a {@link SCMSourceTrait}.
+ *
+ * @since 3.4.0
+ */
+public class MercurialInstallationSCMSourceTrait extends SCMSourceTrait {
+    /**
+     * The {@link MercurialInstallation#getName()} or {@code null} to use the "system" default.
+     */
+    private final @CheckForNull String installation;
+
+    /**
+     * Stapler constructor.
+     *
+     * @param installation the {@link MercurialInstallation#getName()} or {@code null} to use the "system" default.
+     */
+    @DataBoundConstructor public MercurialInstallationSCMSourceTrait(@CheckForNull String installation) {
+        this.installation = Util.fixEmpty(installation);
+    }
+
+    /**
+     * Returns the {@link MercurialInstallation#getName()}.
+     *
+     * @return the {@link MercurialInstallation#getName()} or {@code null} to use the "system" default.
+     */
+    public @CheckForNull String getInstallation() {
+        return installation;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override protected <B extends SCMSourceContext<B, R>, R extends SCMSourceRequest> void decorateContext(B context) {
+        ((MercurialSCMSourceContext<?>)context).withInstallation(installation);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override protected <B extends SCMBuilder<B, S>, S extends SCM> void decorateBuilder(B builder) {
+        ((MercurialSCMBuilder<?>) builder).withInstallation(installation);
+    }
+
+    /**
+     * Our {@link hudson.model.Descriptor}
+     */
+    @Extension public static class DescriptorImpl extends SCMSourceTraitDescriptor {
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override public @Nonnull String getDisplayName() {
+            return "Select Mercurial installation";
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override public boolean isApplicableToBuilder(@Nonnull Class<? extends SCMBuilder> builderClass) {
+            if (super.isApplicableToBuilder(builderClass) && MercurialSCMBuilder.class.isAssignableFrom(builderClass)) {
+                for (MercurialInstallation i : MercurialInstallation.allInstallations()) {
+                    if (i.isUseCaches()) {
```

Comment:
Well…if you have no installations set to `useCaches` then you cannot use `MercurialSCMSource` at all. So whenever this method is called, it should be returning true.